File: src/main/java/chat/teco/tecochat/member/application/MemberService.java
package chat.teco.tecochat.member.application;

import chat.teco.tecochat.member.application.dto.SignUpCommand;
import chat.teco.tecochat.member.domain.Member;
import chat.teco.tecochat.member.domain.MemberRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@RequiredArgsConstructor
@Transactional
@Service
public class MemberService {

    private final MemberRepository memberRepository;

    public void signUp(SignUpCommand signUpCommand) {
        Member member = signUpCommand.toDomain();
        memberRepository.findByName(member.name())
                .ifPresentOrElse(
                        saved -> saved.changeCourse(member.course()),
                        () -> memberRepository.save(member)
                );
    }
}

File: src/main/kotlin/chat/teco/tecochat/application/MemberService.kt
package chat.teco.tecochat.application

import chat.teco.tecochat.domain.member.MemberRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Transactional
@Service
class MemberService(
    private val memberRepository: MemberRepository
) {

    fun signUp(memberData: MemberData) {
        val member = memberData.toMember()

        memberRepository.findByName(member.name())?.apply {
            changeCourse(member.course())
        } ?: run {
            memberRepository.save(member)
        }
    }
}
